src/imagej/__init__.py

```diff
@@ -1251,15 +1253,13 @@ def show_gui_and_run_callbacks(ij):
                     raise
         else:
             # Create and show the application.
-            gateway = _create_gateway()
-            show_gui_and_run_callbacks(gateway)
+            gateway = show_gui_and_run_callbacks()
             # We are responsible for our own blocking.
             # TODO: Poll using something better than ui().isVisible().
             while gateway.ui().isVisible():
                 time.sleep(1)
 
-        del gateway
```

Comment:
This commit deserves to have a more detailed commit message, explaining why it's OK not to `del gateway` after the UI is no longer visible. (I understand that the `del gateway`, where it was previously placed, is probably a bug, but this commit does away with it in *all* cases... are we sure that is what we want?)

Reply:
I'm inclined to revert this to it's previous state where we `del gateway` and `return None` instead of returning the `gateway` when the UI is no longer visible. In gui mode it makes sense to me to assume the user is done with ImageJ when they close the UI, thus deleting the `gateway` and returning `None` makes sense to me.

@ctrueden From what I can tell you added `del gatway` in this commit: fec5ceee42ff1991b50b5e3e3178bc31d16f1339 and it hasn't moved. It doesn't look like a bug to me, but maybe you see something I don't?

Reply:
The reason I said "probably a bug" in that placement is because the `gateway` variable is assigned *inside the else clause*, but deleted *outside the else clause*, meaning it might be deleted even though it was never assigned. If you `del` something that isn't in the globals or locals, Python would issue a `NameError: name 'gateway' is not defined`. But I am not 100% confident about this specific case, since I didn't test the code. I assumed @gselzer had a reason for changing it, which he simply did not write down, but he later said he does not remember, so... testing needed.

Reply:
🙈

Reply:
Thanks Curtis for the breakdown. I see what you mean. I'll see if I can test the condition you laid out and will report back.

Reply:
You're correct @ctrueden in that if you hit the `del gateway` block without assigning it you get a `NameError: name 'gateway' is not defined` error. But I think a simple fix here is just to create the global variable and assign it with `None`. In my testing, this resolves the `NameError` and still retains the gateway cleanup behavior we want.

Reply:
I believe I resolved this and one additional bug here: 45c1448795a69a3cc6db4c6291eebe3e601c5e26 with a detailed explanation.